dali/operators/image/crop/bbox_crop.cc

```diff
@@ -14,143 +14,649 @@
-    bounding_boxes.emplace_back(box);
+    const auto &selected_boxes  = prospective_crop.boxes;
+    WriteBoxesToOutput(ws, selected_boxes);
+
+    if (has_labels_) {
+      const auto &selected_labels = prospective_crop.labels;
+      DALI_ENFORCE(selected_boxes.size() == selected_labels.size(),
+        make_string("Expected boxes.size() == labels.size(). Received: ", selected_boxes.size(),
+          " != ", selected_labels.size()));
+      WriteLabelsToOutput(ws, selected_labels);
+    }
   }
 
-  const auto &labels_tensor = ws.Input<CPUBackend>(1);
+ private:
+  struct ProspectiveCrop {
+    bool success = false;
+    BoundingBox<ndim> crop{};
+    std::vector<BoundingBox<ndim>> boxes;
+    std::vector<int> labels;
+
+    ProspectiveCrop(bool success, const BoundingBox<ndim> &crop, const std::vector<BoundingBox<ndim>> &boxes,
+                    const std::vector<int> &labels)
+        : success(success), crop(crop), boxes(boxes), labels(labels) {}
+    ProspectiveCrop() = default;
+  };
+
+  const ProspectiveCrop FindProspectiveCrop(const std::vector<BoundingBox<ndim>> &bounding_boxes,
+                                            const std::vector<int> &labels,
+                                            int sample) {
+    auto &rng = rngs_[sample];
+    std::uniform_int_distribution<> idx_dist(0, sample_options_.size() - 1);
+    SampleOption option = sample_options_[idx_dist(rng)];
+
+    assert(has_crop_shape_ == has_input_shape_);
+    bool absolute_crop_dims = has_crop_shape_;
+
+    if (option.no_crop) {
+      BoundingBox<ndim> no_crop;
+      for (int d = 0; d < ndim; d++) {
+        no_crop[d] = 0.0f;
+        no_crop[ndim + d] = 1.0f;
+      }
+      if (absolute_crop_dims) {
+        auto &input_shape = input_shape_[sample];
+        for (int d = 0; d < ndim; d++)
+          no_crop[ndim + d] *= input_shape[d];
+      }
+      return ProspectiveCrop(true, no_crop, bounding_boxes, labels);
+    }
 
-  std::vector<int> labels;
-  labels.reserve(static_cast<size_t>(labels_tensor.dim(0)));
+    std::array<float, coords_size> shape, anchor, out_bounds, rel_bounds;
+    for (int i = 0; i < num_attempts_; i++) {
+      if (absolute_crop_dims) {
+        auto &crop_shape = crop_shape_[sample];
+        auto &input_shape = input_shape_[sample];
 
-  for (int i = 0; i < labels_tensor.dim(0); ++i) {
-    const auto *label_data = labels_tensor.data<int>() + i;
-    labels.emplace_back(*label_data);
+        DALI_ENFORCE(crop_shape.sample_dim() == ndim,
+                     make_string("Unexpected number of dimensions. Expected ", ndim, ", got ",
+                                 crop_shape.sample_dim()));
+
+        for (int d = 0; d < ndim; d++) {
+          shape[d] = static_cast<float>(crop_shape[d]);
+          out_bounds[ndim + d] = shape[d];
+          rel_bounds[ndim + d] = shape[d] / input_shape[d];
+        }
+
+        for (int d = 0; d < ndim; d++) {
+          assert(input_shape[d] >= crop_shape[d]);
```

Comment:
This is rather DALI_ENFORCE as it is wrong user input. Maybe in such case we should pad the image and adjust bboxes as well?